net/xia/ppal_ether/main.c

```diff
@@ -81,6 +81,264 @@ static int local_dump_ether(struct fib_xid *fxid, struct fib_xid_table *xtbl,
 	return -EMSGSIZE;
 }
 
+/* ETHER main table operations. */
+static int main_newroute(struct xip_ppal_ctx *ctx, struct fib_xid_table *xtbl,
+			 struct xia_fib_config *cfg)
+{
+	struct interface_addr *neigh_addr, *exist_addr;
+	struct xip_deferred_negdep_flush *dnf;
+	struct fib_xid_ether_main *mether;
+	struct net_device *out_interface;
+	struct fib_xid *cur_fxid;
+	u32 nl_flags, bucket;
+	const char *id;
+	int rc;
+
+	/* Check for errors in cfg. */
+	if (!cfg->xfc_dst || !cfg->xfc_odev || !cfg->xfc_lladdr 
+		        || cfg->xfc_lladdr_len != cfg->xfc_odev->addr_len )
+		return -EINVAL;
+	
+	out_interface = cfg->xfc_odev;
+	nl_flags      = cfg->xfc_nlflags;
+	id            = cfg->xfc_dst->xid_id;
+
+	if (!(out_interface->flags & IFF_UP) || (out_interface->flags & IFF_LOOPBACK))
+		return -EINVAL;
+
+	neigh_addr = allocate_interface_addr(out_interface , cfg->xfc_lladdr , GFP_ATOMIC);
+	if (!neigh_addr)
+		return -ENOMEM;
+	cur_fxid = ether_rt_iops->fxid_find_lock(&bucket, xtbl, id);
+
+	if (cur_fxid)
+	{
+		/* Found a matching fxid. */
+		if (cur_fxid->fx_table_id != XRTABLE_MAIN_INDEX) {
+			rc = -EINVAL;
+			goto unlock_bucket;
+		}
+		mether = fxid_mether(cur_fxid);
+
+		if (cmp_addr(mether->neigh_addr, neigh_addr->ha, neigh_addr->outgress_interface))
+		{
+			/* Exact duplicate entry request. */
+			if ( (nl_flags & NLM_F_EXCL) || !(nl_flags & NLM_F_REPLACE) ) 
+			{
+				rc = -EEXIST;
+				goto unlock_bucket;
+			}
+			rc = 0;
+			goto unlock_bucket;
+		}
+
+		if (mether->neigh_addr)
+		{
+			/* Already allocated neighbour-address case. */
+			if (!(nl_flags & NLM_F_REPLACE)){
+				rc = -EEXIST;
+				goto unlock_bucket;
+			}
+			exist_addr = mether->neigh_addr;
+			del_interface_addr(exist_addr);
+			free_interface_addr(exist_addr);
+		}
+		else
+		{
+			if (!(nl_flags & NLM_F_CREATE)) {
+				rc = -ENOENT;
+				goto unlock_bucket;
+			}
+		}
+
+		rc = attach_neigh_addr_to_fib_entry(mether , neigh_addr);
+		ether_rt_iops->fib_unlock(xtbl, &bucket);
+		if (rc)
+			goto free_addr;
+		return 0;
+	}
+
+	if (!(nl_flags & NLM_F_CREATE)) {
+		rc = -ENOENT;
+		goto unlock_bucket;
+	}
+
+	dnf = fib_alloc_dnf(GFP_ATOMIC);
+	if (!dnf) {
+		rc = -ENOMEM;
+		goto unlock_bucket;
+	}
+
+	mether = ether_rt_iops->fxid_ppal_alloc(sizeof(*mether),GFP_ATOMIC);
```

Comment:
`main_newroute()` is called in process context since it called via Netlink, so we don't need to use `GFP_ATOMIC` here. Please use `GFP_KERNEL` instead.